### PR TITLE
Core - Implement Extended and NFT output objects

### LINKED ISSUE #201: Core - Implement new output objects
New output types:
- [x] Extended output
~~- [ ] Alias output~~
~~- [ ] Foundry output~~
- [x] NFT output

### PR DESCRIPTION
**Pull request by matjazv**
# Description of change

Fixes #201 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Unit tests are in files test_output_extended.c and test_output_nft.c

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests using CTest that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/core/models/outputs/output_extended.c [hunk 1]
```diff
@@ -1,4 +1,300 @@
...
+  offset += address_serialize(output->address, offset, address_serialized_len(output->address));
+
+  // amount
+  memcpy(offset, &output->amount, sizeof(uint64_t));
+  offset += sizeof(uint64_t);
+
+  // native tokens
+  if (output->native_tokens) {
+    offset +=
+        native_tokens_serialize(&output->native_tokens, offset, native_tokens_serialize_len(&output->native_tokens));
```
Comment (oopsmonk) [inline src/core/models/outputs/output_extended.c:172]: Native tokens must be in lexicographical order based on TokenID, therefore `native_tokens_serialize` should do the sorting that can guarantee serialized data is valid.
let's file an issuer to fix it, #223

File: src/core/models/outputs/output_extended.c [hunk 2]
```diff
@@ -1,4 +1,300 @@
...
+  } else {
+    memset(offset, 0, sizeof(uint16_t));
+    offset += sizeof(uint16_t);
+  }
+
+  // feature blocks
+  if (output->feature_blocks) {
+    offset +=
+        feat_blk_list_serialize(output->feature_blocks, offset, feat_blk_list_serialize_len(output->feature_blocks));
```
Comment (oopsmonk) [inline src/core/models/outputs/output_extended.c:181]: the serialized data is valid if and only if block types are in ascending order. file an issue to fix it, #224

File: src/core/models/outputs/output_extended.c [hunk 3]
```diff
@@ -1,4 +1,300 @@
...
+#include "utlist.h"
+
+output_extended_t* output_extended_new(address_t* addr, uint64_t amount, native_tokens_t* tokens,
+                                       feat_blk_list_t* feat_blocks) {
+  if (addr == NULL) {
+    printf("[%s:%d] invalid parameters\n", __func__, __LINE__);
+    return NULL;
+  }
+
+  if (amount < 1000000) {
```
Comment (oopsmonk) [inline src/core/models/outputs/output_extended.c]: please give an identifier for this constant, ex `#define MIN_DUST_ALLOWANCE 1000000`

File: tests/core/test_output_nft.c [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+  address_t addr = {};
+  addr.type = ADDRESS_TYPE_NFT;
+  iota_crypto_randombytes(addr.address, ADDRESS_NFT_BYTES);
+
+  // create NFT ID
+  byte_t nft_id[ADDRESS_NFT_BYTES];
+  memcpy(nft_id, addr.address, ADDRESS_NFT_BYTES);
+
+  // create metadata
+  byte_buf_t* metadata = byte_buf_new_with_data("Test metadata...", sizeof("Test metadata..."));
```
Comment (oopsmonk) [inline tests/core/test_output_nft.c]: please fix the following warnings
```
[build] ../tests/core/test_output_nft.c:41:49: warning: passing 'char [17]' to parameter of type 'byte_t *' (aka 'unsigned char *') converts between pointers to integer types with different sign [-Wpointer-sign]
[build]   byte_buf_t* metadata = byte_buf_new_with_data("Test metadata...", sizeof("Test metadata..."));
[build]                                                 ^~~~~~~~~~~~~~~~~~
[build] ../src/core/utils/byte_buffer.h:40:43: note: passing argument to parameter 'data' here
[build] byte_buf_t* byte_buf_new_with_data(byte_t data[], size_t len);
[build]                                           ^
[build] ../tests/core/test_output_nft.c:150:49: warning: passing 'char [17]' to parameter of type 'byte_t *' (aka 'unsigned char *') converts between pointers to integer types with different sign [-Wpointer-sign]
[build]   byte_buf_t* metadata = byte_buf_new_with_data("Test metadata...", sizeof("Test metadata..."));
[build]                                                 ^~~~~~~~~~~~~~~~~~
[build] ../src/core/utils/byte_buffer.h:40:43: note: passing argument to parameter 'data' here
[build] byte_buf_t* byte_buf_new_with_data(byte_t data[], size_t len);
[build]                                           ^
[build] ../tests/core/test_output_nft.c:365:49: warning: passing 'char [17]' to parameter of type 'byte_t *' (aka 'unsigned char *') converts between pointers to integer types with different sign [-Wpointer-sign]
[build]   byte_buf_t* metadata = byte_buf_new_with_data("Test metadata...", sizeof("Test metadata..."));
[build]                                                 ^~~~~~~~~~~~~~~~~~
[build] ../src/core/utils/byte_buffer.h:40:43: note: passing argument to parameter 'data' here
[build] byte_buf_t* byte_buf_new_with_data(byte_t data[], size_t len);
[build]                                           ^
[build] 3 warnings generated.
```

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: tests/core/test_output_nft.c (hunk 4 above)
Comment (matjazv) [reply tests/core/test_output_nft.c]: I didn't check compilation with Clang but only with GCC. For the future, both are needed to be used before creating PR.

**Review by rebinmannady93 — APPROVED** [review]

**Review by oopsmonk — APPROVED** [review]
lgfm

**Pull request merged**